DeprecatedFile: Migrate "resolve_executable_from_environment" from LibFileSystem to LibCore

BenWiederhake opened the pull request:
This PR answers #18804 with "No, LibCore should not depend on LibFileSystem, instead let's move that one function to Core::System".

Since the change in UserspaceEmulator is so trivial, and because it doesn't compile anyway (see #8872), it is fully migrated away from DeprecatedFile in this PR. Untangling does not seem helpful. I also tried to keep the diff as contained as possible, to reduce interference with efforts to bring UE back to life.

Advances #17129.
Closes #18804.

~~Conflicts #19087, sadly in an unavoidable way:~~
- ~~#19087 changes a comment from `Core::DeprecatedFile::resolve_executable_from_environment` to `FileSystem::resolve_executable_from_environment`~~
- ~~This PR wants to change that comment to `Core::System::resolve_executable_from_environment`~~
- ~~No matter which PR gets merged first, the other has to change it again.~~

~~However, Both PRs can already be independently reviewed and (one of them) merged.~~

EDIT: PR is now ready.

BenWiederhake commented:
Changes since last push: Rebased, now that #19087 is merged.

BenWiederhake commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -1655,4 +1656,39 @@ ErrorOr<void> posix_fallocate(int fd, off_t offset, off_t length)
+// This constant is copied from LibFileSystem. We cannot use or even include it directly,
+// because that would cause a dependency of LibCore on LibFileSystem, effectly rendering
+// the distinction between these libraries moot.
+static constexpr StringView INTERNAL_DEFAULT_PATH_SV = "/usr/local/sbin:/usr/local/bin:/usr/bin:/bin"sv;
```
We don't have LibFileSystem available in LibCore, and including a header risks breaking something at runtime. I'd rather copy a constant over than cause a weird Heisenbug that needs to be figured out later.

Yes, I could just use the literal in the function, but I'd prefer naming these magic values, especially since it's a copy of somewhere else.

konradekk replied:
i meant you could remove this constant altogether from here and settle for the following signature:
```cpp
ErrorOr<String> resolve_executable_from_environment(StringView filename, int flags, default_path)
```
and feed it in call-site instead.

BenWiederhake replied:
Why would someone want to change the default_path? I don't see any advantage, and using any other value will introduce corner cases and subtle differences in behavior.

No, I'd rather keep the signature the same.

BenWiederhake commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -1305,14 +1305,15 @@ ErrorOr<void> exec(StringView filename, ReadonlySpan<StringView> arguments, Sear
+            auto maybe_executable = resolve_executable_from_environment(filename_string);
 
-            if (!maybe_executable.has_value()) {
+            if (maybe_executable.is_error()) {
                 errno_rollback.set_override_rollback_value(ENOENT);
                 return Error::from_errno(ENOENT);
             }
```
You're right, thanks!

GMTA commented on Userland/DevTools/UserspaceEmulator/main.cpp:
```diff
@@ -43,15 +44,17 @@ int main(int argc, char** argv, char** env)
+        executable_path_or_error = FileSystem::real_path(arguments[0]);
+    } else {
+        executable_path_or_error = FileSystem::resolve_executable_from_environment(arguments[0]);
+    }
```
Suggestion:
```cpp
auto executable_path_or_error = arguments[0].contains('/')
    ? FileSystem::real_path(arguments[0])
    : FileSystem::resolve_executable_from_environment(arguments[0]);
```

BenWiederhake replied:
Normally I don't like ternaries, but here it avoids the dummy value nicely. Thanks!

BenWiederhake commented:
Changes since last push:
- Rebased for convenience
- Rewrite a weird if-with-dummy-init into a ternary. One of the few cases where a ternary is easier and better than an if! Thanks @GMTA

BenWiederhake commented:
CI failed:
```
CRASHED LibC/TestMath (466ms)
         Test:   TestMath (crashed)
mmap writable: Bad address
ASSERTION FAILED: false
Userland/Libraries/LibELF/DynamicLoader.cpp:503
```
However, this PR doesn't touch any relevant code, so I'll assume it's just a flake.

Changes since last push:
- no-op rebase to restart CI

timschumi commented on Userland/DevTools/UserspaceEmulator/main.cpp:
```diff
@@ -43,15 +44,14 @@ int main(int argc, char** argv, char** env)
-        executable_path = Core::DeprecatedFile::resolve_executable_from_environment(arguments[0]).value_or({});
-    if (executable_path.is_empty()) {
+    auto executable_path_or_error = arguments[0].contains('/')
+        ? FileSystem::real_path(arguments[0])
```
`resolve_executable_from_environment` already checks if the path contains a `/`, maybe we should make it translate the path into an absolute one instead? It certainly looks like something that other callers of that function might want as well. Having a single non-resolving branch in a resolve function is certainly weird.

BenWiederhake replied:
I really want to modify UserspaceEmulator as little as possible, because it doesn't currently compile and I don't want to conflict with PRs that might bring it back to life.

Whoever fixes UserspaceEmulator can definitely cange this invocation however they see fit, but I don't think it should be changed in this PR.

timschumi commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -1655,4 +1655,39 @@ ErrorOr<void> posix_fallocate(int fd, off_t offset, off_t length)
 }
 #endif
 
+// This constant is copied from LibFileSystem. We cannot use or even include it directly,
+// because that would cause a dependency of LibCore on LibFileSystem, effectly rendering
+// the distinction between these libraries moot.
```
I _think_ the original reason for the distinction was that we didn't really want to create yet another sub-namespace inside `LibCore`, or at least some people were vehemently against it. I'd have to look it up in the original PR again.

BenWiederhake replied:
Looks like it was changed away from a sub-namespace only close to the end: https://github.com/SerenityOS/serenity/pull/17789#pullrequestreview-1350262636

Anyway, I agree that I don't want to re-open that can of worms. That's why I suggest to move this one function back to LibCore, and keep the rest in LibFileSystem.

We could also "officially" move DEFAULT_PATH and DEFAULT_PATH_SV to LibCore, but that raises the question which namespace to use. Do you have an idea where it could go, or is this okay for now?

GMTA commented on Userland/Libraries/LibCore/System.cpp:
```diff
@@ -1655,4 +1656,39 @@ ErrorOr<void> posix_fallocate(int fd, off_t offset, off_t length)
 #endif
 
+// This constant is copied from LibFileSystem. We cannot use or even include it directly,
+// because that would cause a dependency of LibCore on LibFileSystem, effectly rendering
```
Typo: `effectly` -> `effectively`

BenWiederhake commented:
Changes since last push:
- Fixed the `effectly` typo (Thanks @GMTA!)
- Rebased for convenience

@timschumi: I'd like to keep the `INTERNAL_DEFAULT_PATH_SV` as-is for now, because I want to avoid the topic of the LibCore/LibFileSystem split as much as possible. You're very welcome to reorganize it as you see fit, but my goal with this PR is to migrate away from DeprecatedFile.
